src/meta/processors/admin/Balancer.cpp

```diff
@@ -490,50 +614,88 @@ ErrorOr<nebula::cpp2::ErrorCode, bool> Balancer::getHostParts(GraphSpaceID space
   auto properties = MetaServiceUtils::parseSpace(value);
   if (totalParts != properties.get_partition_num()) {
     LOG(ERROR) << "Partition number not equals";
+    LOG(ERROR) << totalParts << " : " << properties.get_partition_num();
     return false;
   }
 
-  if (properties.group_name_ref().has_value()) {
+  int32_t replica = properties.get_replica_factor();
+  LOG(INFO) << "Replica " << replica;
+  if (dependentOnGroup && properties.group_name_ref().has_value()) {
     auto groupName = *properties.group_name_ref();
-    if (dependentOnGroup) {
-      auto zonePartsRet = assembleZoneParts(groupName, hostParts);
-      if (zonePartsRet != nebula::cpp2::ErrorCode::SUCCEEDED) {
-        LOG(ERROR) << "Assemble Zone Parts failed group: " << groupName;
-        return zonePartsRet;
-      }
+    auto groupKey = MetaServiceUtils::groupKey(groupName);
+    std::string groupValue;
+    retCode = kv_->get(kDefaultSpaceId, kDefaultPartId, groupKey, &groupValue);
+    if (retCode != nebula::cpp2::ErrorCode::SUCCEEDED) {
+      LOG(ERROR) << "Get group " << groupName
+                 << " failed: " << apache::thrift::util::enumNameSafe(retCode);
+      return retCode;
+    }
+
+    int32_t zoneSize = MetaServiceUtils::parseZoneNames(std::move(groupValue)).size();
+    LOG(INFO) << "Zone Size " << zoneSize;
+    innerBalance_ = (replica == zoneSize);
+
+    auto activeHostsRet = ActiveHostsMan::getActiveHostsWithGroup(kv_, spaceId);
+    if (!nebula::ok(activeHostsRet)) {
+      return nebula::error(activeHostsRet);
+    }
+
+    std::vector<HostAddr> expand;
+    auto activeHosts = nebula::value(activeHostsRet);
+    std::vector<HostAddr> lostHosts;
+    calDiff(hostParts, activeHosts, expand, lostHosts);
+    // confirmedHostParts is new part allocation map after balance, it would include newlyAdded
+    // and exclude lostHosts
+    HostParts confirmedHostParts(hostParts);
+    for (const auto& h : expand) {
+      LOG(INFO) << "Found new host " << h;
+      confirmedHostParts.emplace(h, std::vector<PartitionID>());
+    }
+    for (const auto& h : lostHosts) {
+      LOG(INFO) << "Lost host " << h;
+      confirmedHostParts.erase(h);
+    }
+
+    auto zonePartsRet = assembleZoneParts(groupName, confirmedHostParts);
+    if (zonePartsRet != nebula::cpp2::ErrorCode::SUCCEEDED) {
+      LOG(ERROR) << "Assemble Zone Parts failed group: " << groupName;
+      return zonePartsRet;
```

Comment:
What is this? I guess you just wanna generate `innerBalance_`

Reply:
`assembleZoneParts` collect some info about zone

Reply:
Explain what are you doing from 638 to 663. Why we `calDiff` and `assembleZoneParts`, nothing is returned.

Reply:
if the space dependent on group, it use to record the partition contained in the zone related to the node.

`assembleZoneParts ` save it in relatedParts_.

calDiff save it in the argument which is reference .